agent/bench-scripts/pbench-fio

```diff
@@ -575,7 +575,9 @@ function fio_run_job() {
 		for client in `echo $clients | sed -e s/,/" "/g`; do
 			client_opts="$client_opts --client=$client"
 			if [ ! -z $numjobs ]; then
-				let max_jobs=$numjobs
+                # so that a separate filename will be generated
+                # for each (host, job-number) pair
+				let max_jobs=$max_jobs+$numjobs
```

Comment:
Isn't `numjobs` a command line parameter?  If it is, won't it default to `1` but potentially some other value that is not one?  If so, how will it work in that case?

Reply:
it defaults to 1 unless you override it.   It's the number of jobs per host, effectively.  See [fio documentation](https://fio.readthedocs.io/en/latest/fio_doc.html) and search for definition there.

Reply:
So are you then looking for `$max_jobs` to be the total number of jobs running across all hosts?

Is so, why not just count up the number of hosts and multiply by `$numjobs`?

Reply:
You could do that, but I was trying to make it easy for readers to see the similarity between the two cases, which are a) no numjobs parameter defined (defaulting to 1), but targets are defined, and b) only 1 target is defined, but numjobs is defined to be > 1 by maximizing the common code between them,   In either case, max_jobs is increased for each client host in the same way.

Reply:
@portante it's not clear why I should change this.

Reply:
This might be a long post, but seems worth to look at this in detail.  The code fragment in question is lines [570](https://github.com/distributed-system-analysis/pbench/pull/1150/files#diff-aa16cf8d4d3d68ccbdda0a744df85bdcR578) - [588](https://github.com/distributed-system-analysis/pbench/pull/1150/files#diff-aa16cf8d4d3d68ccbdda0a744df85bdcR588).

The sole purpose of this code fragment appears to be to fill in the `$client_opts` variable for use when the `fio.cmd` file is created at line [591](https://github.com/distributed-system-analysis/pbench/pull/1150/files#diff-aa16cf8d4d3d68ccbdda0a744df85bdcR591).  The variable is only filled in if there is a `$client_file` parameter or a `$clients` parameter.

If either parameter exists, in addition to adding the `--client` arguments, a `--max-jobs` argument is added.  The calculation of the value to use for the `--max-jobs` parameter is needs to be sorted out.

In the case where a `$client_file` parameter exists, the `$client_file` is given as the argument to one `--client=` parameter, and the value given to the `--max-jobs=` parameter is simply calculated to be the number of clients specified in the `$client_file`, regardless of the value of `$numjobs`.

In the case where a `$clients` parameter exists, one `--client=` parameter is added with a single client value pulled from the `$clients` list.  As each client value is considered, the value to use for the `--max-jobs=` parameter is calculated in two different ways:

 1. When `$numjobs` exists, that value is added to `$max_jobs` each time, effectively (# of clients) * (`$numjobs`)
 2. When `$numjobs` does not exist, the number of `$targets` is used as the value of `$numjobs`, which is effectively (# of clients) * (`$numjobs`)

It seems like this code fails to properly account for `$client_file` as a result.

Perhaps we can consider the following instead for this logic:

```
if [ ! -z $numjobs ]; then
    # so that a separate filename will be generated
    # for each (host, job-number) pair
    local local_numjobs=$numjobs
else
    local local_numjobs=0
    for t in `echo $targets | sed -e s/,/" "/g` ; do
        let local_numjobs=$local_numjobs+1
    done
fi
if [ ! -z "$client_file" ] ;then
    typeset -i nclients=$(wc -l $client_file | cut -d ' ' -f 1)
    local max_jobs=$nclients*$local_numjobs
    client_opts="--client=$client_file --max-jobs=$max_jobs"
elif [ ! -z "$clients" ]; then
    local nclients=0
    for client in `echo $clients | sed -e s/,/" "/g`; do
        client_opts="$client_opts --client=$client"
        let nclients=$nclients+1
    done
    local max_jobs=$nclients*$local_numjobs
    client_opts="$client_opts --max-jobs=$max_jobs"
fi
```

Finally, the question remains if this calculation is correct when **_not_** using the `directory` mode in the job file.

Reply:
used tabs instead of spaces to make it more readable.      I disagree with, or perhaps don't fully understand, your analysis, but I can't change it as you requested.    Your change goes from 1 line of code to 20, but it's not clear to me what it buys you.    My goal is not to fix pbench-fio for everyone, only to add distributed filesystem testing capability.  This 1-line change does that.

Reply:
specifically, if there is only 1 client, the code reduces to exactly what it was before the change, since max_jobs is initialized to zero.

Reply:
> used tabs instead of spaces to make it more readable

The point of the code fragment was to make sure it fit compactly in the window.  Using 8 space tabs in the original file would be great for the comment lines, as well.

> it's not clear to me what it buys you

First, if you use a clients file instead of listing clients on the `pbench-fio` command line, it does not seem your change would work.

The point of the 20 lines of code change is to make sure the same behavior is consistent across the different ways clients are provided without repeating code (DRY principle).

Second, it seems that this behavior is tied to using `directory` in a job file, so the outstanding question that sill has to be answered is if this change works when NOT use `directory`.

Reply:
I think the new code is consistent with the different ways clients are provided.  Also, I tested with an fio run on RBD block devices not using directory in a job file, and that correctly ran as well.  So I believe this is resolved.